test/utils/FunctionCallGenerator.sol

```diff
@@ -0,0 +1,160 @@
+pragma solidity ^0.8.23;
+
+import {console2} from "forge-std/console2.sol";
+import {Test} from "forge-std/Test.sol";
+import {EnumerableSetLib} from "solady/utils/EnumerableSetLib.sol";
+import {ERC20} from "openzeppelin-contracts/contracts/token/ERC20/ERC20.sol";
+import {ERC20Mock} from "openzeppelin-contracts/contracts/mocks/token/ERC20Mock.sol";
+import {IKeyManagement} from "../../src/interfaces/IKeyManagement.sol";
+import {IERC7821} from "../../src/interfaces/IERC7821.sol";
+import {TestKeyManager, TestKey} from "./TestKeyManager.sol";
+import {Call} from "../../src/libraries/CallLib.sol";
+import {Key, KeyLib, KeyType} from "../../src/libraries/KeyLib.sol";
+import {Settings, SettingsLib} from "../../src/libraries/SettingsLib.sol";
+import {HandlerCall, CallUtils} from "./CallUtils.sol";
+import {ExecuteHandler} from "./ExecuteHandler.sol";
+import {IInvariantStateTracker, InvariantStateTracker} from "./InvariantStateTracker.sol";
+import {IMinimalDelegation} from "../../src/interfaces/IMinimalDelegation.sol";
+
+/**
+ * @title FunctionCallGenerator
+ * @dev Helper contract to generate random function calls for MinimalDelegation invariant testing
+ */
+abstract contract FunctionCallGenerator is Test, InvariantStateTracker {
+    using EnumerableSetLib for EnumerableSetLib.Bytes32Set;
+    using KeyLib for Key;
+    using CallUtils for Call;
+    using CallUtils for Call[];
+    using CallUtils for HandlerCall;
+    using CallUtils for HandlerCall[];
+    using TestKeyManager for TestKey;
+
+    uint256 public constant FUZZED_FUNCTION_COUNT = 3;
+
+    uint256 public constant MAX_DEPTH = 5;
+    uint256 public constant MAX_KEYS = 10;
+
+    IMinimalDelegation internal immutable signerAccount;
+    address private immutable _tokenA;
+    address private immutable _tokenB;
+
+    // Keys that will be operated over in generated calldata
+    TestKey[] public fixture_testKeys;
+
+    constructor(IMinimalDelegation _signerAccount, address tokenA, address tokenB) {
+        signerAccount = _signerAccount;
+        _tokenA = tokenA;
+        _tokenB = tokenB;
+
+        // Generate MAX_KEYS and add to fixtureKey
+        for (uint256 i = 0; i < MAX_KEYS; i++) {
+            fixture_testKeys.push(TestKeyManager.withSeed(KeyType.Secp256k1, vm.randomUint()));
+        }
+    }
+
+    function _rand(TestKey[] storage keys, uint256 seed) internal view returns (TestKey memory) {
+        return keys[seed % keys.length];
+    }
+
+    function _wrapCallFailedRevertData(bytes4 selector) internal pure returns (bytes memory) {
+        return abi.encodeWithSelector(IERC7821.CallFailed.selector, abi.encodePacked(selector));
+    }
+
+    /// @return calldata to register a new key along with its callback
+    function _registerCall(TestKey memory newKey, bytes memory revertData)
+        internal
+        virtual
+        returns (HandlerCall memory)
+    {
+        // No error is thrown if the key is already registered, so ignore
```

Comment:
there is an error thrown though if you try to register a key that is the signer account

Reply:
i believe that won't happen here since the newKey is randomly generated though right?

Reply:
(also same for update call)

Reply:
Ah yeah I guess you would need a collision

Reply:
i think maybe add a case in the case that we end up combining key state and the root key is an option as an input into these function generators

Reply:
added in new PR